nexus/inventory/src/builder.rs

```diff
@@ -769,6 +990,28 @@ mod test {
         assert!(collection
             .caboose_for(CabooseWhich::RotSlotB, &sled1_bb)
             .is_none());
+        let rot_page =
+            collection.rot_page_for(RotPageWhich::Cmpa, &sled1_bb).unwrap();
+        assert!(collection.rot_pages.contains(&rot_page.page));
+
+        // TODO-correctness Is this test correct? We reported the same RoT page
+        // with different data (rot_page1, then rot_page2). The second
+        // `found_rot_page` returned an error, but we overwrote the original
+        // data and did not record the error in `collection.errors`. Should we
+        // either have kept the original data or returned Ok while returning an
+        // error? It seems a little strange we returned Err but accepted the new
+        // data.
+        assert_eq!(rot_page.page.data_base64, rot_page2.data_base64);
```

Comment:
This is the second test issue I mentioned in the PR description. The caboose behaves this way so I made it match, but I'm not sure it's right - should we at least be logging a collection error if we hit this?

Reply:
If I'm understanding it right: I think the current behavior is okay and also the improvements you alluded to would be good, too.  My thinking is this:

- It seems very possible (although probably uncommon) to get two different responses here due to the underlying data having changed in between the two calls (e.g., if we did an inventory collection during an update, we might see the caboose contents change).  In this case, there's an intrinsic race, and I don't think it matters much which data we take, especially as long as we record which one it was (source and timestamp).  A consumer that was depending on getting the later data really needs to wait for a collection to complete that _starts_ after whatever operation they want it to be after.  (I realize in the current implementation we'll take the later one anyway.  Similarly if someone was depending on it being the earlier one, they'd have to look at a collection that finished before whatever operation they care about happened.)
- The builder returns an error so that the caller can at least tell that something unusual happened and can report it.  What level of reporting is appropriate?  I think calling this a collection error is misleading.  Those are intended to communicate that a particular collection is likely to be incomplete or invalid, which isn't the case in the example I mentioned and I don't think we want that sort of race to invalidate any concurrent collection.  I think a warning-level log message (and counter bump, ideally) would be appropriate.  If we wanted to augment collections with warnings or something like that, I think that'd also be fine, but I'm not sure it's worth doing right now.
- The other possibility is that two MGS's reliably report inconsistent data for the same device even though the underlying data isn't changing.  I can imagine this happening (i.e., due to a bug in one of them), but dealing with it in general seems pretty hard and it doesn't seem all that likely to me.  Being able to tell in a support context that it's happened seems like the right level of effort.

That's just an explanation of how I think we got here.  We could certainly change any of this!

Reply:
This makes sense, thanks. I was only thinking of your third bullet point (or slight variations, such as an SP / RoT reporting inconsistent data itself, even if MGS isn't buggy), but you're right that this seems we're much more likely to hit this by racing with an update, which shouldn't cause a failure.

> I think a warning-level log message (and counter bump, ideally) would be appropriate.

`Collector` does log this (at `error!`) level currently - want me to change them to `warn!`, and possibly include a "was an update happening simultaneously?" (either in the log message itself or in a comment around where the log message is generated)?

What do you mean by a counter bump?

Reply:
I wouldn't bother changing it.

By counter bump I mean adding Oximeter-collected metrics to this subsystem and having an error counter for this case.  That'd be welcome but obviously beyond the scope of this PR!